Refactoring of Phase1 OMTF emulator code

mbluj opened the pull request:
#### PR description:

In this PR code of OMTF L1Mu emulator is refactored and (partly) cleaned. It is needed as preparation for 2024 data-taking with an updated algorithm that enables triggering with non-pointing muons, and to decouple Phase1 (current) and Phase2 emulators.

Note: this PR does not switch on algorithm changes expected for 2024, but only prepares code for it. This will be done with other PR when new algorithm in firmware is fully validated. It will additional require an upload of new patters to DB and updated GT or (temporary) new xml files in cms-data. 

#### PR validation:

Successfully passed matrix tests (limited)

FYI, @akalinow, @kbunkow

mbluj committed:
Refactoring and cleaning of Phase1 OMTF emulator code

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"0497014b52e4afd679909ad3850916dd5d61c9ca":{"files":["L1Trigger/L1TMuonOverlapPhase1/interface/AlgoMuonBase.h","L1Trigger/L1TMuonOverlapPhase1/interface/AngleConverterBase.h","L1Trigger/L1TMuonOverlapPhase1/interface/MuonStub.h","L1Trigger/L1TMuonOverlapPhase1/interface/MuonStubMakerBase.h","L1Trigger/L1TMuonOverlapPhase1/interface/Omtf/AlgoMuon.h","L1Trigger/L1TMuonOverlapPhase1/interface/Omtf/GhostBuster.h","L1Trigger/L1TMuonOverlapPhase1/interface/Omtf/GhostBusterPreferRefDt.h","L1Trigger/L1TMuonOverlapPhase1/interface/Omtf/GoldenPatternBase.h","L1Trigger/L1TMuonOverlapPhase1/interface/Omtf/GoldenPatternResult.h","L1Trigger/L1TMuonOverlapPhase1/interface/Omtf/GoldenPatternWithStat.h","L1Trigger/L1TMuonOverlapPhase1/interface/Omtf/IOMTFEmulationObserver.h","L1Trigger/L1TMuonOverlapPhase1/interface/Omtf/IProcessorEmulator.h","L1Trigger/L1TMuonOverlapPhase1/interface/Omtf/OMTFConfiguration.h","L1Trigger/L1TMuonOverlapPhase1/interface/Omtf/OMTFProcessor.h","L1Trigger/L1TMuonOverlapPhase1/interface/Omtf/OMTFinput.h","L1Trigger/L1TMuonOverlapPhase1/interface/Omtf/OmtfAngleConverter.h","L1Trigger/L1TMuonOverlapPhase1/interface/Omtf/OmtfName.h","L1Trigger/L1TMuonOverlapPhase1/interface/Omtf/PtAssignmentBase.h","L1Trigger/L1TMuonOverlapPhase1/interface/Omtf/XMLEventWriter.h","L1Trigger/L1TMuonOverlapPhase1/interface/ProcConfigurationBase.h","L1Trigger/L1TMuonOverlapPhase1/interface/Tools/CandidateSimMuonMatcher.h","L1Trigger/L1TMuonOverlapPhase1/interface/Tools/DataROOTDumper2.h","L1Trigger/L1TMuonOverlapPhase1/interface/Tools/EmulationObserverBase.h","L1Trigger/L1TMuonOverlapPhase1/interface/Tools/PatternGenerator.h","L1Trigger/L1TMuonOverlapPhase1/interface/Tools/PatternOptimizerBase.h","L1Trigger/L1TMuonOverlapPhase1/plugins/L1TMuonOverlapPhase1TrackProducer.cc","L1Trigger/L1TMuonOverlapPhase1/python/fakeOmtfParams_cff.py","L1Trigger/L1TMuonOverlapPhase1/python/simOmtfDigis_cfi.py","L1Trigger/L1TMuonOverlapPhase1/python/simOmtfDigis_extrapolSimple_cfi.py","L1Trigger/L1TMuonOverlapPhase1/src/AlgoMuonBase.cc","L1Trigger/L1TMuonOverlapPhase1/src/AngleConverterBase.cc","L1Trigger/L1TMuonOverlapPhase1/src/MuonStub.cc","L1Trigger/L1TMuonOverlapPhase1/src/MuonStubMakerBase.cc","L1Trigger/L1TMuonOverlapPhase1/src/Omtf/AlgoMuon.cc","L1Trigger/L1TMuonOverlapPhase1/src/Omtf/GhostBuster.cc","L1Trigger/L1TMuonOverlapPhase1/src/Omtf/GhostBusterPreferRefDt.cc","L1Trigger/L1TMuonOverlapPhase1/src/Omtf/GoldenPatternBase.cc","L1Trigger/L1TMuonOverlapPhase1/src/Omtf/GoldenPatternResult.cc","L1Trigger/L1TMuonOverlapPhase1/src/Omtf/OMTFConfiguration.cc","L1Trigger/L1TMuonOverlapPhase1/src/Omtf/OMTFProcessor.cc","L1Trigger/L1TMuonOverlapPhase1/src/Omtf/OMTFReconstruction.cc","L1Trigger/L1TMuonOverlapPhase1/src/Omtf/OMTFSorter.cc","L1Trigger/L1TMuonOverlapPhase1/src/Omtf/OMTFinput.cc","L1Trigger/L1TMuonOverlapPhase1/src/Omtf/OMTFinputMaker.cc","L1Trigger/L1TMuonOverlapPhase1/src/Omtf/OmtfAngleConverter.cc","L1Trigger/L1TMuonOverlapPhase1/src/Omtf/OmtfName.cc","L1Trigger/L1TMuonOverlapPhase1/src/Omtf/ProcessorBase.cc","L1Trigger/L1TMuonOverlapPhase1/src/Omtf/XMLConfigReader.cc","L1Trigger/L1TMuonOverlapPhase1/src/Omtf/XMLConfigWriter.cc","L1Trigger/L1TMuonOverlapPhase1/src/Omtf/XMLEventWriter.cc","L1Trigger/L1TMuonOverlapPhase1/src/ProcConfigurationBase.cc","L1Trigger/L1TMuonOverlapPhase1/src/Tools/CandidateSimMuonMatcher.cc","L1Trigger/L1TMuonOverlapPhase1/src/Tools/DataROOTDumper2.cc","L1Trigger/L1TMuonOverlapPhase1/src/Tools/EmulationObserverBase.cc","L1Trigger/L1TMuonOverlapPhase1/src/Tools/EventCapture.cc","L1Trigger/L1TMuonOverlapPhase1/src/Tools/OMTFConfigMaker.cc","L1Trigger/L1TMuonOverlapPhase1/src/Tools/PatternGenerator.cc","L1Trigger/L1TMuonOverlapPhase1/src/Tools/PatternOptimizerBase.cc","L1Trigger/L1TMuonOverlapPhase1/src/Tools/StubsSimHitsMatching.cc","L1Trigger/L1TMuonOverlapPhase1/test/expert/omtf/modifyHwToLogixLayerXml.py","L1Trigger/L1TMuonOverlapPhase1/test/expert/omtf/runMuonOverlapDataDump.py","L1Trigger/L1TMuonOverlapPhase1/test/expert/omtf/runMuonOverlapPatternGenerator.py","L1Trigger/L1TMuonOverlapPhase1/test/expert/omtf/runMuonOverlapPatternGeneratorClassProb.py","L1Trigger/L1TMuonOverlapPhase1/test/expert/omtf/runMuonOverlapPatternGenerator_displ.py","L1Trigger/L1TMuonOverlapPhase1/test/expert/omtf/runMuonOverlap_Pats0x0013_efficiencyAnalyser.py","L1Trigger/L1TMuonOverlapPhase1/test/expert/omtf/runMuonOverlap_Pats0x0013_efficiencyAnalyser_displMu.py","L1Trigger/L1TMuonOverlapPhase1/test/expert/omtf/runMuonOverlap_Pats0x0013_generateExtrFactors.py","L1Trigger/L1TMuonOverlapPhase1/test/expert/omtf/runMuonOverlap_newPats_efficiencyAnalyser.py","L1Trigger/L1TMuonOverlapPhase1/test/runMuonOverlap_2023withExtrapol_mc.py","L1Trigger/L1TMuonOverlapPhase1/test/runMuonOverlap_run2_data.py","L1Trigger/L1TMuonOverlapPhase1/test/runMuonOverlap_run3_data.py","L1Trigger/L1TMuonOverlapPhase1/test/runMuonOverlap_run3_mc.py"],"time":1704826765},"emoji":{}} -->

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43687/38368

- This PR adds an extra 152KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43687/38368/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43687/38368/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

mbluj committed:
code-checks

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43687/38369

- This PR adds an extra 156KB to repository

cmsbuild commented:
A new Pull Request was created by  @mbluj  for master.

It involves the following packages:

- L1Trigger/L1TMuonOverlapPhase1 (**l1**)


@epalencia, @cmsbuild, @aloeliger can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @dinyar, @eyigitba, @thomreis, @JanFSchulte, @missirol this is something you requested to watch as well.
@antoniovilela, @sextonkennedy, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

epalencia commented:
Please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa87dd/36792/summary.html
**COMMIT**: d016cbb4de293f7c07e8abca13123400a54e8212
**CMSSW**: CMSSW_14_0_X_2024-01-10-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43687/36792/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 150 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-01-10-1100+fa87dd/60537/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 26 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-01-10-1100+fa87dd/60537/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3247277
 - DQMHistoTests: [Total failures: 1280](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-01-10-1100+fa87dd/60537/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3245975
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 200 log files, 161 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-01-10-1100+fa87dd/60537/triggerResults): no differences found

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_14_1_X. Please open a backport if it should also go in to CMSSW_14_0_X.

aloeliger commented on L1Trigger/L1TMuonOverlapPhase1/src/Omtf/GoldenPatternResult.cc:
```diff
@@ -316,6 +321,123 @@ void GoldenPatternResult::finalise9() {
+        } else {
+          firedLayerBits &= ~(1 << iLogicLayer);
+          stubResults[iLogicLayer].setValid(false);
+          //if(stubResults[iLogicLayer].getPdfVal() == 0) pdfSum -= 64;; //there was hit, but it did not fire to the pdf - this is not possible here, since the banding layer if fired here
+          //so in this case simply:
+          //pdfSum += 0;
+        }
+      } else {
+        //bending layer fired, but not fits to the pdf, N.B works only with the patterns having "no hit value" and with noHitValueInPdf = True
+        /*if (stubResults[iLogicLayer].getPdfVal() == 0)
+          pdfSum -= 32; //has no sense with extrapolation from the ref layer using the phiB
+        else*/
+        pdfSum += stubResults[iLogicLayer].getPdfVal();  //bending layer not fired at all
+      }
+    } else {
+      if (iLogicLayer < 10 && stubResults[iLogicLayer].getPdfVal() == 0)
+        pdfSum -= 32;
```
Would you please remove some of the commented code unless it serves a demonstrative purpose, and maybe leave a comment explaining where `10` and `32` are coming from here?

aloeliger commented on L1Trigger/L1TMuonOverlapPhase1/src/Omtf/OMTFProcessor.cc:
There are a number of places in this file with commented code and a number of constants without explanation of units or where it came from. Would you please remove the commented code unless it serves to demonstrate something, and leave a comment for the units of things like `rTargetLayer`, just in case other L1 developers have to look at this code for maintenance or reference?

mbluj replied:
set of comments added

aloeliger commented on L1Trigger/L1TMuonOverlapPhase1/src/Omtf/OMTFSorter.cc:
```diff
@@ -35,29 +36,41 @@ AlgoMuons::value_type OMTFSorter<GoldenPatternType>::sortRefHitResults(
+      }
+    }
+
+    //    LogTrace("OMTFReconstruction")<<"OMTFSorter::sortRefHitResults() " << itGP->key()<<" getFiredLayer "<< itGP->getResults()[procIndx][iRefHit].getFiredLayerCnt()
+    //    <<" PdfSum "<< itGP->getResults()[procIndx][iRefHit].getPdfSum()
+    //    <<" PdfSumUnconstr "<< itGP->getResults()[procIndx][iRefHit].getPdfSumUnconstr()<< std::endl;
```
```suggestion
```

aloeliger commented on L1Trigger/L1TMuonOverlapPhase1/src/Omtf/OMTFinput.cc:
```diff
@@ -47,22 +49,44 @@ const int OMTFinput::getHitEta(unsigned int iLayer, unsigned int iInput) const {
 }
 
+const int OMTFinput::getHitQual(unsigned int iLayer, unsigned int iInput) const {
+  /*  assert(iLayer < muonStubsInLayers.size());
+  assert(iInput < muonStubsInLayers[iLayer].size());*/
```
```suggestion
```

aloeliger commented on L1Trigger/L1TMuonOverlapPhase1/src/Omtf/OMTFinputMaker.cc:
```diff
@@ -93,11 +92,15 @@ void CscDigiToStubsConverterOmtf::addCSCstubs(MuonStubPtrs2D& muonStubsInLayers,
   unsigned int iLayer = config->getHwToLogicLayer().at(hwNumber);
   unsigned int iInput = OMTFinputMaker::getInputNumber(config, rawid, iProcessor, procTyp);
 
+  //edm::LogVerbatim("l1tOmtfEventPrint")<<"addCSCstubs iProcessor "<<iProcessor<<" procTyp "<<procTyp<< std::endl;
```
```suggestion
```

aloeliger commented on L1Trigger/L1TMuonOverlapPhase1/src/Omtf/OmtfAngleConverter.cc:
```diff
@@ -150,6 +108,25 @@ int OmtfAngleConverter::getGlobalEta(const DTChamberId dTChamberId,
   //const DTChamberId dTChamberId(aDigi.whNum(),aDigi.stNum(),aDigi.scNum()+1);
   DTTrigGeom trig_geom(_geodt->chamber(dTChamberId), false);
 
+  /* debug printout to check the geometry of the chambers
+  Local2DPoint chamberMiddleLP(0, 0);
+  GlobalPoint chamberMiddleGP = _geodt->chamber(dTChamberId)->toGlobal(chamberMiddleLP);
+  float phin = (dTChamberId.sector()-1)*Geom::pi()/6;
+  float phiRF = _geodt->chamber(dTChamberId)->position().phi();
+  float deltaPhi = phiRF - phin;
+
+  LogTrace("l1tOmtfEventPrint")<<"OmtfAngleConverter::getGlobalEta "<<dTChamberId
+    <<" perp "<<chamberMiddleGP.perp()
+    //<<" chamber()->position().perp() "<<_geodt->chamber(dTChamberId)->position().perp()
+    <<" x "<<_geodt->chamber(dTChamberId)->position().x()
+    <<" y "<<_geodt->chamber(dTChamberId)->position().y()
+    <<" z "<<_geodt->chamber(dTChamberId)->position().z()
+    <<" - phiRF "<<phiRF << " rad "<< phiRF * 180. / M_PI<<" deg "
+    <<" - phin "<<phin<< " rad "<< phin * 180. / M_PI<<" deg "
+    <<" - deltaPhi "<<deltaPhi<<" r "<<chamberMiddleGP.perp() * cos(deltaPhi);
+    //<<" distSL "<<trig_geom.distSL();
+  */
+
```
```suggestion
```

aloeliger commented on L1Trigger/L1TMuonOverlapPhase1/src/Omtf/OmtfAngleConverter.cc:
```diff
@@ -241,22 +224,58 @@ int OmtfAngleConverter::getGlobalEta(unsigned int rawid, const CSCCorrelatedLCTD
+  r = final_gp.perp();
 
-  return etaKeyWG2Code(id, keyWG);
+  //edm::LogVerbatim("l1tOmtfEventPrint")<<id<<" ETA_CSC st: " << id.station()<< "ri: "<<id.ring()<<" eta: " <<  final_gp.eta()
+  //           <<" etaCode_simple: " <<  etaVal2Code( final_gp.eta() )<< " keyWG: "<<keyWG <<" etaKeyWG2Code: "<<etaKeyWG2Code(id,keyWG)<< std::endl;
```
```suggestion
```

aloeliger commented on L1Trigger/L1TMuonOverlapPhase1/src/Omtf/OmtfAngleConverter.cc:
```diff
@@ -241,22 +224,58 @@ int OmtfAngleConverter::getGlobalEta(unsigned int rawid, const CSCCorrelatedLCTD
+    const LocalPoint lpWg = layer_geom->localCenterOfWireGroup(keyWG);
+    const GlobalPoint gpWg = layer->surface().toGlobal(lpWg);
+
+    //edm::LogVerbatim("l1tOmtfEventPrint")<<"gpWg.eta() " << gpWg.eta()<<" etaToHwEta(gpWg.eta()) "<<config->etaToHwEta(gpWg.eta()) << std::endl;
```
```suggestion
```

aloeliger commented on L1Trigger/L1TMuonOverlapPhase1/src/Omtf/OmtfAngleConverter.cc:
```diff
@@ -241,22 +224,58 @@ int OmtfAngleConverter::getGlobalEta(unsigned int rawid, const CSCCorrelatedLCTD
   const GlobalPoint gp = roll->toGlobal(lp);
 
-  return etaVal2Code(gp.eta());
+  if (id.region() == 0) {  //barrel
+    /* //debug printout to check the geometry of the chambers
+    float phin = (id.sector()-1)*Geom::pi()/6;
+    float phiHit = gp.phi();
+    float deltaPhi = phiHit - phin;
+
+    LogTrace("l1tOmtfEventPrint")<<"OmtfAngleConverter::getGlobalEtaRpc "<<id
+        <<" perp "<<gp.perp()
+        //<<" chamber()->position().perp() "<<_geodt->chamber(dTChamberId)->position().perp()
+        <<" x "<<gp.x()
+        <<" y "<<gp.y()
+        <<" z "<<gp.z()
+        <<" - phiRF "<<phiHit << " rad "<< phiHit * 180. / M_PI<<" deg "
+        <<" - phin "<<phin<< " rad "<< phin * 180. / M_PI<<" deg "
+        <<" - deltaPhi "<<deltaPhi<<" r "<<gp.perp() * cos(deltaPhi);
+        */
+  } else {
+    r = gp.perp();
+  }
```
```suggestion
  if (id.region() != 0)
    r = gp.perp();
```

aloeliger commented on L1Trigger/L1TMuonOverlapPhase1/src/Omtf/XMLConfigReader.cc:
```diff
@@ -401,7 +401,7 @@ std::unique_ptr<GoldenPatternType> XMLConfigReader::buildGP(DOMElement *aGPEleme
     std::string strVal = _toString(aItemElement->getAttribute(xmlTresh));
     thresholds[iItem] = std::stof(strVal);
-    //LogTrace("l1tOmtfEventPrint")<<__FUNCTION__<<":"<<__LINE__<<" strVal "<<strVal<<" thresholds[iItem] "<<thresholds[iItem]<<std::endl;
+    //LogTrace("OMTFReconstruction")<<__FUNCTION__<<":"<<__LINE__<<" strVal "<<strVal<<" thresholds[iItem] "<<thresholds[iItem]<<std::endl;
```
```suggestion
```

aloeliger commented on L1Trigger/L1TMuonOverlapPhase1/src/Tools/CandidateSimMuonMatcher.cc:
```diff
@@ -101,8 +104,8 @@ bool simTrackIsMuonInOmtf(const SimTrack& simTrack) {
   //otherwise many candidates are marked as ghosts
-  //if( (fabs(simTrack.momentum().eta()) >= 0.82 ) && (fabs(simTrack.momentum().eta()) <= 1.24) ) {
-  if ((fabs(simTrack.momentum().eta()) >= 0.72) && (fabs(simTrack.momentum().eta()) <= 1.3)) {
+  //if( (std::abs(simTrack.momentum().eta()) >= 0.82 ) && (std::abs(simTrack.momentum().eta()) <= 1.24) ) {
```
```suggestion
```
I can't really tell whether this file was designed to work in general CMSSW operation, so for the next few comments, I think you can ignore some of the picky stuff from me if it isn't.

mbluj replied:
It is indeed not designed in general CMSSW operation, i.e. in standard emulation workflow. It is a tool which matches simulated muon with OMTF response (from emulator) used for test or preparing training samples for NN. 
Anyway, thank you for your comments we will try to address them to improve quality of this code.

aloeliger commented on L1Trigger/L1TMuonOverlapPhase1/src/Tools/CandidateSimMuonMatcher.cc:
```diff
@@ -272,6 +313,7 @@ std::vector<const l1t::RegionalMuonCand*> CandidateSimMuonMatcher::ghostBust(
 }
 
 TrajectoryStateOnSurface CandidateSimMuonMatcher::atStation2(FreeTrajectoryState ftsStart, float eta) const {
+  eta = 0;  //fix me!!!!! in case of displaced muon the vertex eta has no sense
```
Should this be fixed?

aloeliger commented on L1Trigger/L1TMuonOverlapPhase1/src/Tools/CandidateSimMuonMatcher.cc:
```diff
@@ -575,6 +619,18 @@ std::vector<MatchingResult> CandidateSimMuonMatcher::match(std::vector<const l1t
       continue;  //no sense to do matching
     }
 
+    //checking if the propagated track is inside the OMTF range, TODO - tune the range!!!!!!!!!!!!!!!!!
+    //eta 0.7 is the beginning of the MB2,
+    //the eta range wider than the nominal OMTF region is needed, as in any case muons outside this region are seen by the OMTF
+    //so it better to train the nn suich that is able to measure its pt, as it may affect the rate
+    if ((std::abs(tsof.globalPosition().eta()) >= 0.7) && (std::abs(tsof.globalPosition().eta()) <= 1.31)) {
+      LogTrace("l1tOmtfEventPrint")
```
Would you mind updating this with an explanation of `1.31` as well?

aloeliger commented on L1Trigger/L1TMuonOverlapPhase1/src/Tools/CandidateSimMuonMatcher.cc:
```diff
@@ -637,13 +701,27 @@ std::vector<MatchingResult> CandidateSimMuonMatcher::match(
+    //eta 0.7 is the beginning of the MB2,
+    if ((std::abs(tsof.globalPosition().eta()) >= 0.7) && (std::abs(tsof.globalPosition().eta()) <= 1.3)) {
+      LogTrace("l1tOmtfEventPrint")
+          << "CandidateSimMuonMatcher::match trackingParticle IS in OMTF region, matching to the omtfCands";
```
Previous numbers had this as `1.31`. Given the comment this looks like it is describing the same region? Should this be `1.3` or `1.31`? Could you please confirm and update the comment to reflect what that number represents?

aloeliger commented on L1Trigger/L1TMuonOverlapPhase1/src/Tools/EventCapture.cc:
```diff
@@ -96,13 +97,15 @@ void EventCapture::observeEventEnd(const edm::Event& iEvent,
-      if (matchingResult.muonCand && matchingResult.muonCand->hwQual() >= 12 &&
-          matchingResult.muonCand->hwPt() > 38) {  //&& matchingResult.genPt < 20
+      if (matchingResult.muonCand)
+      //&& matchingResult.muonCand->hwQual() >= 12 &&
+      //matchingResult.muonCand->hwPt() > 38)
+      {  //&& matchingResult.genPt < 20
```
```suggestion
      { 
```

aloeliger commented on L1Trigger/L1TMuonOverlapPhase1/src/Tools/PatternGenerator.cc:
Going by the looks of this this is also used for firmware creation/testing so I personally won't ask for anything, but would recommend just cleaning up the commented code.

aloeliger commented on L1Trigger/L1TMuonOverlapPhase1/test/expert/omtf/modifyHwToLogixLayerXml.py:
```diff
@@ -0,0 +1,142 @@
+        
+    ########################################################################################    
+        
+    # Write back to the XML file
+    out_xml_file = '/afs/cern.ch/work/k/kbunkow/public/CMSSW/cmssw_13_x_x/CMSSW_13_1_0/src/L1Trigger/L1TMuon/data/omtf_config/hwToLogicLayer_0x0209.xml'
+    #tree.write(out_xml_file, encoding='utf-8', xml_declaration=True)
+    
+    formatted_xml = minidom.parseString(ET.tostring(root)).toprettyxml(indent="  ")
+    formatted_xml = os.linesep.join([s for s in formatted_xml.splitlines() if s.strip()])
+    with open(out_xml_file, 'w', encoding='utf-8') as output_file:
+        output_file.write(formatted_xml)
+    
+    print("out_xml_file ", out_xml_file)
+
+if __name__ == "__main__":
+    xml_file = '/afs/cern.ch/work/k/kbunkow/public/CMSSW/cmssw_13_x_x/CMSSW_13_1_0/src/L1Trigger/L1TMuon/data/omtf_config/hwToLogicLayer_0x0209org.xml'
+    modify_xml(xml_file)
```
This seems somewhat specific to Karol. Should this be in CMSSW?

aloeliger commented on L1Trigger/L1TMuonOverlapPhase1/test/expert/omtf/runMuonOverlapPatternGenerator_displ.py:
This file also seems to contain a few things that are very specific to Karol. Should this file be in CMSSW?

aloeliger commented on L1Trigger/L1TMuonOverlapPhase1/test/expert/omtf/runMuonOverlap_Pats0x0013_efficiencyAnalyser.py:
This file also contains some user specific references, and I'm going to assume quite a few the files after it might as well. Do the user specific files need to be in CMSSW?

aloeliger reviewed (commented):
Apologies for the long time before review. I have gone through. I think some things here are duplicated from the conversation on the L1 page, but some of this should be new.

kbunkow commented on L1Trigger/L1TMuonOverlapPhase1/src/Omtf/OMTFProcessor.cc:
```diff
@@ -424,6 +822,92 @@ void OMTFProcessor<GoldenPatternType>::printInfo() const {
   ProcessorBase<GoldenPatternType>::printInfo();
 }
 
+template <class GoldenPatternType>
+void OMTFProcessor<GoldenPatternType>::saveExtrapolFactors() {
+  //if(this->myOmtfConfig->nProcessors() == 3) //phase2
+  extrapolMultiplier = 512;
+
+  boost::property_tree::ptree tree;
+  auto& extrFactorsTree = tree.add("ExtrapolationFactors", "");
+  extrFactorsTree.add("<xmlattr>.multiplier", extrapolMultiplier);
```
extrapolMultiplier is needed to convert floating point values to integer values, as in the phase-1 firmware the arithmetical operations are performed with the use of the integer types (see OMTFProcessor<GoldenPatternType>::extrapolateDtPhiBFixedPoint() ). The method saveExtrapolFactors() is used to generate the extrapolation factors, which are then stored in the XML file. It is used only in a special job when the flag useFloatingPointExtrapolation is set in the python. For the "normal" operation of the emulator, the extrapolation factors are loaded from this file, together with the extrapolMultiplier.  Thus the class-default value of the 128 is always overwitten at the job configuration stage. 
The value of 512 which is in saveExtrapolFactors() is just devoted to one particular version of the firmware (phase-2 in this case).

These parameters should go to the CMSSW database, but the needed updates are not yet ready.

kbunkow commented on L1Trigger/L1TMuonOverlapPhase1/interface/Omtf/OMTFinput.h:
```diff
@@ -27,6 +29,7 @@ class OMTFinput : public MuonStubsInput {
   ///Method used in DiMuon studies.
   void mergeData(const OMTFinput *aInput);
 
+  //TODOO make it: if the layer is bending layer, the phiB from the iLayer -1 is returned, change in all places it is used
```
This comment points to a possible improvement (clean up) to the codes, which would allow avoiding "if's" like e.g. in [ProcessorBase.cc#L156](https://github.com/cms-sw/cmssw/blob/c4e19155815230c34f92d11ab87d8f34db0cfbac/L1Trigger/L1TMuonOverlapPhase1/src/Omtf/ProcessorBase.cc#L156)
However it would have to be done carefully, so it is postponed... but the TODO reminds us about it.

kbunkow commented on L1Trigger/L1TMuonOverlapPhase1/src/Omtf/XMLEventWriter.cc:
```diff
@@ -12,15 +12,47 @@
 XMLEventWriter::~XMLEventWriter() {}
 
+void XMLEventWriter::observeProcesorBegin(unsigned int iProcessor, l1t::tftype mtfType) {
+  if (eventNum > 5000)
+    return;
```
Yes, the XMLEvenWriter is used for dumping the data for the firmware testing. 
If I remember correctly there was a problem here: the XML elements were stored in the memory, so the job was consuming more and more memory and was crashing finally without any meaningful information. To avoid this cut `eventNum > 5000` was hard coded.

But it was when we used xercesc::DOM here, later we moved to boost/property_tree for the XML creation, maybe it is smarter, we should check if this cut is still needed.

aloeliger commented:
This PR requires https://github.com/cms-data/L1Trigger-L1TMuon/pull/26

smuzaffar commented:
test parameter:
- pull_request = https://github.com/cms-data/L1Trigger-L1TMuon/pull/26

smuzaffar commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa87dd/37487/summary.html
**COMMIT**: d016cbb4de293f7c07e8abca13123400a54e8212
**CMSSW**: CMSSW_14_1_X_2024-02-14-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/43687/37487/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 14 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-14-2300+fa87dd/61123/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 69 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-14-2300+fa87dd/61123/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3248554
 - DQMHistoTests: [Total failures: 1259](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-14-2300+fa87dd/61123/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3247273
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 200 log files, 161 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-14-2300+fa87dd/61123/triggerResults): no differences found

mbluj committed:
cleanup in L1Trigger/L1TMuonOverlapPhase1

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43687/38895

- This PR adds an extra 44KB to repository

cmsbuild commented:
Pull request #43687 was updated. @aloeliger, @cmsbuild, @epalencia can you please check and sign again.

mbluj commented:
@aloeliger it is commit with part of changes introduced, it will be completed soon. I will let you know when we are ready.

cmsbuild commented:
Pull request #43687 was updated. @epalencia, @aloeliger, @cmsbuild can you please check and sign again.

mbluj commented:
@aloeliger The last few commits with requested code cleanup and fixes are now squashed to one to not pollute history. I hope it addresses your comments and review can proceed.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43687/38943

- This PR adds an extra 280KB to repository

- Found files with invalid states:
  - L1Trigger/L1TMuonOverlapPhase1/test/expert/omtf/runMuonOverlap_Pats0x0013_efficiencyAnalyser.py:
    - Added: 0497014b52e4afd679909ad3850916dd5d61c9ca
    - Deleted: 728847ff78b6b4c99a9efdb158ef30f3de0ae6b8
  - L1Trigger/L1TMuonOverlapPhase1/test/expert/omtf/runMuonOverlap_Pats0x0013_efficiencyAnalyser_displMu.py:
    - Added: 0497014b52e4afd679909ad3850916dd5d61c9ca
    - Deleted: 728847ff78b6b4c99a9efdb158ef30f3de0ae6b8
  - L1Trigger/L1TMuonOverlapPhase1/test/expert/omtf/modifyHwToLogixLayerXml.py:
    - Added: 0497014b52e4afd679909ad3850916dd5d61c9ca
    - Deleted: 728847ff78b6b4c99a9efdb158ef30f3de0ae6b8
  - L1Trigger/L1TMuonOverlapPhase1/test/expert/omtf/runMuonOverlap_newPats_efficiencyAnalyser.py:
    - Added: 0497014b52e4afd679909ad3850916dd5d61c9ca
    - Deleted: 728847ff78b6b4c99a9efdb158ef30f3de0ae6b8

cmsbuild commented:
Pull request #43687 was updated. @epalencia, @cmsbuild, @aloeliger can you please check and sign again.

mbluj committed:
Update propagation to station 2

aloeliger commented:
please test with https://github.com/cms-data/L1Trigger-L1TMuon/pull/26

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43687/39059

- This PR adds an extra 28KB to repository

- Found files with invalid states:
  - L1Trigger/L1TMuonOverlapPhase1/test/expert/omtf/runMuonOverlap_Pats0x0013_efficiencyAnalyser.py:
    - Added: 0497014b52e4afd679909ad3850916dd5d61c9ca
    - Deleted: 728847ff78b6b4c99a9efdb158ef30f3de0ae6b8
  - L1Trigger/L1TMuonOverlapPhase1/test/expert/omtf/runMuonOverlap_Pats0x0013_efficiencyAnalyser_displMu.py:
    - Added: 0497014b52e4afd679909ad3850916dd5d61c9ca
    - Deleted: 728847ff78b6b4c99a9efdb158ef30f3de0ae6b8
  - L1Trigger/L1TMuonOverlapPhase1/test/expert/omtf/modifyHwToLogixLayerXml.py:
    - Added: 0497014b52e4afd679909ad3850916dd5d61c9ca
    - Deleted: 728847ff78b6b4c99a9efdb158ef30f3de0ae6b8
  - L1Trigger/L1TMuonOverlapPhase1/test/expert/omtf/runMuonOverlap_newPats_efficiencyAnalyser.py:
    - Added: 0497014b52e4afd679909ad3850916dd5d61c9ca
    - Deleted: 728847ff78b6b4c99a9efdb158ef30f3de0ae6b8

cmsbuild commented:
Pull request #43687 was updated. @epalencia, @aloeliger can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa87dd/37693/summary.html
**COMMIT**: e7de87b2c9ab51671b3c564b151eb42815032c2b
**CMSSW**: CMSSW_14_1_X_2024-02-25-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/43687/37693/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @epalencia cms-sw/cmsdist#9029

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa87dd/37693/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa87dd/37693/git-merge-result


## Comparison Summary



**Summary**:
 - You potentially [added 165 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-25-2300+fa87dd/61334/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 63 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-25-2300+fa87dd/61334/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3245365
 - DQMHistoTests: [Total failures: 1256](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-25-2300+fa87dd/61334/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3244087
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 200 log files, 161 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-25-2300+fa87dd/61334/triggerResults): no differences found

aloeliger commented:
+l1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @sextonkennedy, @antoniovilela, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)
**Notice** This PR was tested with additional Pull Request(s), please also merge them if necessary: cms-data/L1Trigger-L1TMuon#26
